### PR TITLE
*: use given task id or random number as task id

### LINKED ISSUE #7484: The unified read pool should not use start_ts as task id
### What version of TiKV are you using?

4.0-beta.2 and up

### What operating system and CPU are you using?

Unrelated

### Steps to reproduce

1. Start a transaction, run a big query (for example `SELECT COUNT(*)` from a big table).
2. Start some other big queries.
3. While those big queries are in process, run a small query (for example a point select) in the transaction started in 1.

### What did you expect?

The small query finishes quickly.

### What did happened?

Because we use start_ts as the task id, the priority of the small query is lowered by the previously executed big query, thus it may take some time to finish. I don't think it's a reasonable strategy.

### PR DESCRIPTION
**Pull request by sticnarf**
- [x] Depends on https://github.com/pingcap/kvproto/pull/604

### What problem does this PR solve?

Issue Number: Closes #7484 

Problem Summary:

It's not suitable in all cases to use start ts as the task id for the unified read pool. For example, in the same transaction, a big query is executed and the following point selects are also downgraded. 

### What is changed and how it works?

What's Changed:

https://github.com/pingcap/kvproto/pull/604 adds `task_id` to kv context. We can use it to generate a better task id instead of using start_ts in coprocessor.
Task id tends to be generated in serial. The algorithm is replacing the high 16 bits of start ts (which is not so representative) with the low 16-bits of the given task id.
This PR also changes to let all storage tasks use random number as task id. These tasks tend to end fast so we don't bother to give them a meaningful task id. (Maybe we can do it in the future)

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

Allow using the task ID provided by the client as the identifier used by the unified read pool to avoid unexpected priority lowered by another task in the same transaction .

### CODE REVIEW

File: src/coprocessor/mod.rs [hunk 1]
```diff
@@ -170,4 +171,42 @@ impl ReqContext {
...
+            Some(0) | Some(MAX_TS) | None => thread_rng().next_u64(),
+            Some(start_ts) => start_ts,
+        };
+        let task_id: u64 = self.context.get_task_id();
+        if task_id > 0 {
+            // It is assumed that the lower bits of task IDs in a single transaction
+            // tend to be different. So if task_id is provided, we concatenate the
+            // low 16 bits of the task_id and the high 48 bits of the start_ts to build
+            // the final task id.
+            (task_id << (64 - ID_SHIFT)) | (base & MASK)
```
Comment (youjiali1995) [inline src/coprocessor/mod.rs:189]: Why not use the low 48 bits of the start_ts?

Comment (youjiali1995) [reply src/coprocessor/mod.rs:189]: I think it's more likely to be different.

Comment (youjiali1995) [reply src/coprocessor/mod.rs:189]: Oh, it's low bits... The comment is wrong.

File: src/coprocessor/mod.rs [hunk 2]
```diff
@@ -170,4 +171,42 @@ impl ReqContext {
...
+        const MAX_TS: u64 = u64::max_value();
+        let base = match self.txn_start_ts {
+            Some(0) | Some(MAX_TS) | None => thread_rng().next_u64(),
+            Some(start_ts) => start_ts,
+        };
+        let task_id: u64 = self.context.get_task_id();
+        if task_id > 0 {
+            // It is assumed that the lower bits of task IDs in a single transaction
+            // tend to be different. So if task_id is provided, we concatenate the
+            // low 16 bits of the task_id and the high 48 bits of the start_ts to build
```
Comment (youjiali1995) [inline src/coprocessor/mod.rs:187]:
```suggestion
            // low 16 bits of the task_id and the low 48 bits of the start_ts to build
```

Comment (sticnarf) [reply src/coprocessor/mod.rs]: Fixed.

**Comment by sticnarf** [comment]
@youjiali1995 PTAL again. Thanks!

**Pull request merged**